### PR TITLE
[Concept Entry] AI Recurrent Neural Networks (RNN) 

### LINKED ISSUE #2017: [Concept Entry] AI Recurrent Neural Networks (RNN)
### Reason/inspiration (optional)

We would like a new concept entry for Recurrent Neural Networks (RNN)

The entry should go in a new file in docs/content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md

The entry should
- Describe what a RNN is.
- Describe its relation to AI.

Please refer to the [entry templates](https://github.com/Codecademy/docs/blob/main/documentation/), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Concept Entrty for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [x] Issue is assigned

### PR DESCRIPTION
**Pull request by apropos0**
### Type of Change

- Adding a new concept entry. Closes #2017 

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

**Comment by apropos0** [comment]
#2017

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+---
+Title: 'Recurrent Neural Networks'
+Description: 'Recurrent Neural Network is a type of neural network distinguished by its use of the outputs from previous layers as inputs in the current layer'
```
Comment (bananagrenade) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:3]: Description: 'Recurrent Neural Network is a type of neural network distinguished by storing and re-using outputs from previous layers as additional inputs in the current layer'

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+CatalogContent:
+  - 'machine-learning'
+  - 'paths/machine-learning-engineer'
+---
+
+## What Is A Recurrent Neural Network (RNN)?
+
+**Recurrent Neural Network** (RNN) is a type of neural network distinguished by its use of the outputs from previous layers as inputs in the current layer; in other words, the results from earlier layers are persisted through the network through hidden states.
+
+RNNs apply best in cases where "memory" or "state" is important, this can be sequential or time-based data, where the output is a predictable value in the sequence.
```
Comment (bananagrenade) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:19]: RNNs apply best in cases where the sequence of data is important to the predicted output. In sequential or time-based data sets, RNNs can persist previous outputs as "memories" or "states", and use that to predict the output in the sequence.

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+## How Does an RNN Work?
+
+To conceptually train an RNN, a vector-encoded input is passed from the dataset to the network, and the network computes its output using initially random weights and biases (unlike other neural networks, an RNN keeps the same weight through all its layers).
```
Comment (bananagrenade) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:23]: To train a RNN, a vector-encoded input is passed from the dataset (_a vector-encoded dataset is passed as an input to the network ?_ might be misunderstanding this part) to a network that is initialized with random weights and biases. However, unlike other neural networks, a RNN keeps the same weight and biases throughout all of its layers when computing the output value(s).

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+When the output is computed, it is fed forward to the next layer, which uses the previous result as well as the current input from the dataset to form its output. This process is continued through the rest of the layers in the network until the final output is returned as a result.
```
Comment (bananagrenade) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:24]: Think you should drop in a sentence or two describing the outputs being stored / persisted for re-use in a loop (?). Then the mention of "previous" result might make more sense

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+Using a loss function, we compare the expected and actual results to find the error. The error is backpropagated through time (BPTT) to update the weights and biases and thus train the network until the expected result is produced.
```
Comment (bananagrenade) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:25]: A loss function is used to compare the expected result(s) and actual result(s) to find the error rate of the prediction. The error is then backpropagated through time (BPTT) to so the network can update its weights and biases in each layer. Repeating this process thus trains the network until the expected result is produced.

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+A standard issue when training a typical RNN is the vanishing gradient problem. When attempting to backpropagate and update the weights, we see that inputs used earlier in the layer have a smaller impact on training as the error becomes smaller as it is backpropagated and used in updating the weights. To rectify this, networks like Long Short Term Memory (LSTM) and Gated recurrent unit (GRU) can be used, which, although out of the scope of this doc, are worth looking into, as well as another common issue that can come up: the exploding gradient problem.
```
Comment (bananagrenade) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md]: A common issue when training a RNN...

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
+---
+Title: 'Recurrent Neural Networks'
+Description: 'Recurrent Neural Network is a type of neural network distinguished by storing and re-using the output from previous steps as an additional input in the current step'
```
Comment (SSwiniarski) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:3]:
```suggestion
Description: 'Recurrent Neural Networks are a type of neural network distinguished by storing and re-using the output from previous steps as an additional input in the current step'
```

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  - 'Computer Science'
+Tags:
+  - 'AI'
+  - 'Neural Networks'
+CatalogContent:
+  - 'machine-learning'
+  - 'paths/machine-learning-engineer'
+---
+
+## What Is A Recurrent Neural Network (RNN)?
```
Comment (SSwiniarski) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:15]:
```suggestion
```
Remove this header. In docs we don't use headers above the first paragraph. The paragraph comes right after the title.

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+**Recurrent Neural Network** (RNN) is a type of neural network distinguished by storing and re-using the outputs from previous steps as additional inputs in the current step. RNNs processes sequential data in groups of "timesteps" which represent a slice of the sequence, these timesteps could represent words from a sentence or frames from an audio clip. The network uses the output from previous timesteps which are stored in a "hidden state" or "hidden variable" to compute the output for the current timestep
```
Comment (SSwiniarski) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:17]:
```suggestion
A **Recurrent Neural Network** (RNN) is a type of neural network distinguished by storing and re-using the outputs from previous steps as additional inputs in the current step. RNNs process sequential data in groups of "timesteps" which represent a slice of the sequence.  These timesteps could represent words from a sentence or frames from an audio clip. The network uses the output from previous timesteps which are stored in a "hidden state" or "hidden variable" to compute the output for the current timestep.
```

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+RNNs are used in cases where the sequence of data is important to the predicted output. In sequential or time-based data sets, RNNs use previous outputs as "memories" or "states", and use that to predict the output in the sequence.
+
+## How Does an RNN Work?
+
+The training data is split into time-based sequences, they are then vector-encoded and passed into the RNN as inputs. The RNN is initialized with random weights and biases which remain the same through all timesteps. The network iterates over the sequence to compute the output or hidden state for each timestep.
```
Comment (SSwiniarski) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:23]:
```suggestion
The training data is split into time-based sequences. They are then vector-encoded and passed into the RNN as inputs. The RNN is initialized with random weights and biases which remain the same through all timesteps. The network iterates over the sequence to compute the output or hidden state for each timestep.
```

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+The current hidden state is computed by passing the input for the current time-step, the previous hidden state and the weights and biases for the sequence into a layer with an activation function of choice. The hidden state is then returned as an output and passed to the next timestep. The process is repeated through the rest of the timesteps. Depending on the task, the final output can either be returned as a single vector or can be a combination of output vectors returned in each timestep. For the first timestep the previous hidden state can be initialized as a zero vector, its size will depend on the intended output size.
```
Comment (SSwiniarski) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:25]:
```suggestion
The current hidden state is computed by passing the input for the current time-step, the previous hidden state, and the weights and biases for the sequence into a layer with an activation function of choice. The hidden state is then returned as an output and passed to the next timestep. The process is repeated through the rest of the timesteps. Depending on the task, the final output can either be returned as a single vector or can be a combination of output vectors returned in each timestep. For the first timestep the previous hidden state can be initialized as a zero vector, its size will depend on the intended output size.
```

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+RNNs are generally used to predict data that forms a series; this can be grouped into input-to-output classes. Some examples are:
+
+- Many inputs to many outputs: Speech Recognition, Language Translation, Language modeling.
+
+- Many inputs to one output: Stock prediction, weather prediction, Sentiment analysis.
+
+- One input to many output: Image captioning, Music Generation.
+
+- One input to one output: Image classification.
```
Comment (SSwiniarski) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:41]:
```suggestion
- Many inputs to many outputs: Speech Recognition, Language Translation, Language modeling.
- Many inputs to one output: Stock prediction, weather prediction, Sentiment analysis.
- One input to many output: Image captioning, Music Generation.
- One input to one output: Image classification.
```
Remove blank lines between items in a bulleted list.

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@apropos0 I did a first maintainer review and left a few comments.

**Comment by apropos0** [comment]
Thanks @SSwiniarski , applied suggestions

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  - 'Computer Science'
+Tags:
+  - 'AI'
+  - 'Neural Networks'
+CatalogContent:
+  - 'machine-learning'
+  - 'paths/machine-learning-engineer'
+---
+
+A **Recurrent Neural Network** (RNN) is a type of neural network distinguished by storing and re-using the outputs from previous steps as additional inputs in the current step. RNNs process sequential data in groups of "timesteps" which represent a slice of the sequence. These timesteps could represent words from a sentence or frames from an audio clip. The network uses the output from previous timesteps which are stored in a "hidden state" or "hidden variable" to compute the output for the current timestep.
```
Comment (HishamT) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:15]:
```suggestion
A **Recurrent Neural Network** (RNN) is a type of neural network that stores and re-uses the outputs from previous steps as additional inputs in the current step. RNNs process sequential data in groups of *timesteps* which represent a segment of the sequence. These timesteps could represent words from a sentence or frames from an audio clip. The network uses the output from previous timesteps which are stored in a "hidden state" or "hidden variable" to compute the output for the current timestep.
```

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+## How Does an RNN Work?
+
+The training data is split into time-based sequences. They are then vector-encoded and passed into the RNN as inputs. The RNN is initialized with random weights and biases which remain the same through all timesteps. The network iterates over the sequence to compute the output or hidden state for each timestep.
+
+The current hidden state is computed by passing the input for the current time-step, the previous hidden state, and the weights and biases for the sequence into a layer with an activation function of choice. The hidden state is then returned as an output and passed to the next timestep. The process is repeated through the rest of the timesteps. Depending on the task, the final output can either be returned as a single vector or can be a combination of output vectors returned in each timestep. For the first timestep the previous hidden state can be initialized as a zero vector, its size will depend on the intended output size.
```
Comment (HishamT) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:23]:
```suggestion
The current hidden state is computed by passing the input for the current time-step, the previous hidden state, and the weights and biases for the sequence into a layer with an activation function of choice. The hidden state is then returned as an output and passed to the next timestep. The process is repeated through the rest of the timesteps. Depending on the task, the final output can either be returned as a single vector or can be a combination of output vectors returned in each timestep. For the first timestep, the previous hidden state can be initialized as a zero vector of size dependent on the intended output size.
```

File: content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+A loss function is used to compare the expected result and actual result returned by the model. Using the loss calculated, the loss gradient for each timestep is computed using backpropagation through time (BPTT) and is used to update the weights in order to reduce the loss. Repeating this process trains the network until the expected result is produced.
```
Comment (HishamT) [inline content/ai/concepts/neural-networks/terms/recurrent-neural-networks/recurrent-neural-networks.md:25]: Capitalize Backpropagation Through Time

**Review by HishamT — COMMENTED** [review]
Hey @apropos0, thank you for your entry to Codecademy Docs. I have reviewed the entry and left some comments.

**Comment by apropos0** [comment]
@HishamT Thanks for the review. Applied suggestions. Used _timesteps_ instead, prettier had an issue with the *timesteps* format

**Comment by SSwiniarski** [comment]
@apropos0 Congrats on the new entry! Live link here: https://www.codecademy.com/resources/docs/ai/neural-networks/recurrent-neural-networks

image
